dipy/denoise/localpca.py

```diff
@@ -0,0 +1,118 @@
+import numpy as np
+import scipy as sp
+import scipy.linalg as sla
+from warnings import warn
+
+# Try to get the SVD through direct API to lapack:
+try:
+    from scipy.linalg.lapack import sgesvd as svd
+    svd_args = [1, 0]
+# If you have an older version of scipy, we fall back
+# on the standard scipy SVD API:
+except ImportError:
+    from scipy.linalg import svd
+    svd_args = [False]
+
+
+def localpca(arr, sigma, patch_radius=2, tau_factor=2.3):
+    r"""Local PCA-based denoising of diffusion datasets.
+
+    Parameters
+    ----------
+    arr : 4D array
+        Array of data to be denoised. The dimensions are (X, Y, Z, N), where N
+        are the diffusion gradient directions.
+    sigma : float or 3D array
+        Standard deviation of the noise estimated from the data.
+    patch_radius : int, optional
+        The radius of the local patch to be taken around each voxel (in
+        voxels). Default: 2 (denoise in blocks of 5x5x5 voxels).
+    tau_factor : float, optional
+        Thresholding of PCA eigenvalues is done by nulling out eigenvalues that
+        are smaller than:
+
+        .. math ::
+
+                \tau = (\tau_{factor} \sigma)^2
+
+        Default: 2.3, based on the results described in [Manjon13]_.
+
+    Returns
+    -------
+    denoised_arr : 4D array
+        This is the denoised array of the same size as that of the input data.
+
+    References
+    ----------
+    .. [Manjon13] Manjon JV, Coupe P, Concha L, Buades A, Collins DL (2013)
+                  Diffusion Weighted Image Denoising Using Overcomplete Local
+                  PCA. PLoS ONE 8(9): e73021.
+                  https://doi.org/10.1371/journal.pone.0073021
+    """
+    if not arr.ndim == 4:
+        raise ValueError("PCA denoising can only be performed on 4D arrays.",
+                         arr.shape)
+
+    patch_size = 2 * patch_radius + 1
+
+    if patch_size ** 3 < arr.shape[-1]:
+        e_s = "You asked for PCA denoising with a "
+        e_s += "patch_radius of {0} ".format(patch_radius)
+        e_s += "for data with {0} directions. ".format(arr.shape[-1])
+        e_s += "This would result in an ill-conditioned PCA matrix. "
+        e_s += "Please increase the patch_radius."
+        raise ValueError(e_s)
+
+    if isinstance(sigma, np.ndarray):
+        if not sigma.shape == arr.shape[:-1]:
+            e_s = "You provided a sigma array with a shape"
+            e_s += "{0} for data with".format(sigma.shape)
+            e_s += "shape {0}. Please provide a sigma array".format(arr.shape)
+            e_s += " that matches the spatial dimensions of the data."
+            raise ValueError(e_s)
+
+    tau = np.ones(arr.shape[:-1]) * ((tau_factor * sigma) ** 2)
+    del sigma
+    # declare arrays for theta and thetax
+    theta = np.zeros(arr.shape, dtype=np.float64)
+    thetax = np.zeros(arr.shape, dtype=np.float64)
```

Comment:
To ease up on memory, since these arrays only store data and not intermediate results, they could be float32 and the computations kept up in float64. Currently running the 1.2mm and it is using 3.8 gb of ram, still manageable but that would help cutting on the usage.